scanpy/tests/test_normalization.py

```diff
@@ -56,3 +62,289 @@ def test_normalize_total_view(typ, dtype):
+
+    ### inplace = True ###
+    # modifies the input adata object
+    # PCA on all genes
+    sc.experimental.pp.normalize_pearson_residuals_pca(
+        adata, inplace=True, n_comps_pca=n_comps_pca
+    )
+    # PCA on hvgs only
+    sc.experimental.pp.normalize_pearson_residuals_pca(
+        adata_with_hvgs, inplace=True, n_comps_pca=n_comps_pca
+    )
+    # PCA again on all genes (hvg use supressed)
+    sc.experimental.pp.normalize_pearson_residuals_pca(
+        adata_not_using_hvgs,
+        inplace=True,
+        n_comps_pca=n_comps_pca,
+        use_highly_variable=False,
+    )
+
+    for ad in [adata, adata_with_hvgs, adata_not_using_hvgs]:
+        # check adata_pca keys are complete
+        assert np.all(
+            np.isin(
+                [
+                    'pearson_residuals_normalization',
+                ],
+                list(ad.uns.keys()),
+            )
+        )
+        assert np.all(np.isin(['X_pca'], list(ad.obsm.keys())))
+        # check shapes: adata should always retains original shape
+        assert ad.shape == (n_cells, n_genes)
+        assert ad.obsm['X_pca'].shape == (n_cells, n_comps_pca)
+
+    # check PC shapes to see whether or not HVGs were used for PCA
+    assert adata.uns['pca']['PCs'].shape == (n_genes, n_comps_pca)
+    assert adata_with_hvgs.uns['pca']['PCs'].shape == (
+        n_hvgs,
+        n_comps_pca,
+    )
+    assert adata_not_using_hvgs.uns['pca']['PCs'].shape == (
+        n_genes,
+        n_comps_pca,
+    )
+
+    # test for inplace/outplace
+    for ad_inplace, ad_outplace in zip(
+        [adata_pca, adata_pca_with_hvgs, adata_pca_not_using_hvgs],
+        [adata, adata_with_hvgs, adata_not_using_hvgs],
+    ):
+        np.testing.assert_array_equal(
+            ad_inplace.obsm['X_pca'],
+            ad_outplace.obsm['X_pca'],
+        )
+
+
+@pytest.mark.parametrize(
+    'sparsity_func', [csr_matrix.toarray, csr_matrix], ids=lambda x: x.__name__
+)
+@pytest.mark.parametrize('dtype', ['float32', 'int64'])
+@pytest.mark.parametrize('n_hvgs', [100, 200])
+@pytest.mark.parametrize('n_comps_pca', [30, 50])
+def test_normalize_pearson_residuals_recipe(sparsity_func, dtype, n_hvgs, n_comps_pca):
+    adata = _prepare_pbmc_testdata(sparsity_func, dtype, small=True)
+    n_cells, n_genes = adata.shape
+
+    adata_with_hvgs = adata.copy()
+    sc.experimental.pp.highly_variable_genes(
+        adata_with_hvgs, flavor='pearson_residuals', n_top_genes=n_hvgs
+    )
+
+    ### inplace = False ###
+    # outputs the (potentially hvg-restricted) adata_pca object
+    # PCA on all genes
+    adata_pca, hvg = sc.experimental.pp.recipe_pearson_residuals(
+        adata.copy(), inplace=False, n_comps_pca=n_comps_pca, n_top_genes=n_hvgs
+    )
+
+    # for both cases, check adata_pca keys are complete
+    assert np.all(
+        np.isin(
+            ['pearson_residuals_normalization', 'pca'],
+            list(adata_pca.uns.keys()),
+        )
+    )
```

Comment:
```suggestion
    assert {'pearson_residuals_normalization', 'pca'}.issubset(adata_obs.uns)
```

A little shorter.  Could be worth just writing a little helper function for this that would always return useful error messages.

Reply:
good idea, done!